catkin_ws/src/state_estimation/src/ekf_sensors.py

```diff
@@ -0,0 +1,179 @@
+#!/usr/bin/env python
+
+import rospy
+import math
+import numpy as np
+import scipy.linalg
+
+import autograd.numpy as np  # Thinly-wrapped numpy
+from autograd import grad, jacobian, elementwise_grad
+
+from geometry_msgs.msg import Point, Vector3, Quaternion
+from sensor_msgs.msg import Imu, FluidPressure
+
+from aquadrone_msgs.msg import SubState, MotorControls
+
+from ekf_indices import IDx as IDx
+
+class BaseSensorListener(object):
+    # https://en.wikipedia.org/wiki/Extended_Kalman_filter
+
+    #
+    # Functions Common for Each Sensor; No need to re-implement
+    #
+
+    def __init__(self):
+        self.last_time = rospy.Time.now().to_sec()
+        self.calc_H = jacobian(self.state_to_measurement_h)
+
+    def is_valid(self):
+        return rospy.Time.now().to_sec() - self.last_time < self.get_timeout_sec()
+
+    def get_H(self, x, u):
+        # Jacobian of measurement wrt state (as calculated by get_measurement_z)
+        H = self.calc_H(x, u)
+        H = np.reshape(H, (self.get_p(), x.shape[0]))
+        return H
+
+    #
+    # Functions Required for Each Sensor; Implement
+    #
+
+    def get_timeout_sec(self):
+        # Amount of time to use old measurements for
+        raise NotImplementedError
+
+    def get_p(self):
+        # Number of elements in measurement vector
+        raise NotImplementedError
+
+    def get_measurement_z(self):
+        # Actual measurement from the sensor
+        raise NotImplementedError
+
+    def get_R(self):
+        # Uncertainty matrix of measurements
+        raise NotImplementedError
+
+    def state_to_measurement_h(self, x, u):
+        # Calculate expected sensor readings from current state and inputs
+        # Needs to be autograd-able
+        raise NotImplementedError
+
+
+
+class PressureSensorListener(BaseSensorListener):
+    def __init__(self):
+        super(PressureSensorListener, self).__init__()
+        self.depth_sub = rospy.Subscriber("aquadrone/out/pressure", FluidPressure, self.depth_cb)
+
+        self.z = 0
+        self.var = 1
+
+        self.pressure_offset = 100.0
+        self.g = 9.8
+
+    def get_timeout_sec(self):
+        return 0.1
+
+    def get_p(self):
+        return 1
+
+    def get_measurement_z(self):
+        vec = np.zeros((1,1))
+        vec[0] = self.z
+        return vec
+
+    def get_R(self):
+        # Variance of measurements
+        var = np.zeros((1,1))
+        var[0,0] = self.var
+        return var
+
+    def state_to_measurement_h(self, x, u):
+        return x[IDx.Pz]
+
+    def depth_cb(self, msg):
+        press = msg.fluid_pressure
+        var = msg.variance
```

Comment:
For both the pressure sensor and gyro, the sensor will likely not return variance data. This will likely have to be read off of their data sheets and hardcoded into their respective nodes.

Reply:
Yeah. I'll need to compare the real vs simulated sensors and try to get them to match (the sim depth sensor has variance, and your PR for our depth sensor will be good to help it match).

The IMU node may come with variance data, but I'll need to check that as well.